drivers/adc/adc_gd32.c

```diff
@@ -0,0 +1,403 @@
+	}
+
+	/* Signle conversion mode with regular group. */
+	ADC_RSQ2(cfg->reg) &= ~ADC_RSQX_RSQN;
+	ADC_RSQ2(cfg->reg) = index;
+
+	data->buffer = sequence->buffer;
+
+	adc_context_start_read(&data->ctx, sequence);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+static int adc_gd32_read(const struct device *dev,
+			 const struct adc_sequence *sequence)
+{
+	struct adc_gd32_data *data = dev->data;
+	int error;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	error = adc_gd32_start_read(dev, sequence);
+	adc_context_release(&data->ctx, error);
+
+	return error;
+}
+
+#ifdef CONFIG_ADC_ASYNC
+static int adc_gd32_read_async(const struct device *dev,
+			       const struct adc_sequence *sequence,
+			       struct k_poll_signal *async)
+{
+	struct adc_gd32_data *data = dev->data;
+	int error;
+
+	adc_context_lock(&data->ctx, true, async);
+	error = adc_gd32_start_read(dev, sequence);
+	adc_context_release(&data->ctx, error);
+
+	return error;
+}
+#endif /* CONFIG_ADC_ASYNC */
+
+static struct adc_driver_api adc_gd32_driver_api = {
+	.channel_setup = adc_gd32_channel_setup,
+	.read = adc_gd32_read,
+#ifdef CONFIG_ADC_ASYNC
+	.read_async = adc_gd32_read_async,
+#endif /* CONFIG_ADC_ASYNC */
+};
+
+static int adc_gd32_init(const struct device *dev)
+{
+	struct adc_gd32_data *data = dev->data;
+	const struct adc_gd32_config *cfg = dev->config;
+	int ret;
+
+	data->dev = dev;
+
+	ret = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
+	}
+
+#ifdef CONFIG_SOC_SERIES_GD32F3X0
+	/* Select adc clock source and its prescaler. */
+	rcu_adc_clock_config(cfg->rcu_clock_source);
+#endif
+
+	rcu_periph_clock_enable(cfg->rcu_periph_clock);
+
+#if defined(CONFIG_SOC_SERIES_GD32F403) || \
+	defined(CONFIG_SOC_SERIES_GD32VF103) || \
+	defined(CONFIG_SOC_SERIES_GD32F3X0)
+	/* Set SWRCST as the regular channel external trigger. */
+	ADC_CTL1(cfg->reg) &= ~ADC_CTL1_ETSRC;
+	ADC_CTL1(cfg->reg) |= CTL1_ETSRC(7);
+
+	/* Enable external trigger for regular channel. */
+	ADC_CTL1(cfg->reg) |= ADC_CTL1_ETERC;
+#endif
+
+	/* Enable ADC */
+	ADC_CTL1(cfg->reg) |= ADC_CTL1_ADCON;
+
+	adc_gd32_calibration(cfg);
+
+	cfg->irq_config_func();
+
+	adc_context_unlock_unconditionally(&data->ctx);
+
+	return 0;
+}
+
+#define HANDLE_IRQS(idx)							\
+	static const struct device *dev_##idx = DEVICE_DT_INST_GET(idx);	\
+	const struct adc_gd32_config *cfg_##idx = dev_##idx->config;		\
+										\
+	if (ADC_CTL0(cfg_##idx->reg) & ADC_CTL0_EOCIE) {			\
+		adc_gd32_isr(dev_##idx);					\
+	}
+
+static void adc_gd32_shared_irq_handler(void)
+{
+	DT_INST_FOREACH_STATUS_OKAY(HANDLE_IRQS);
+}
+
+static void adc_gd32_shared_irq_configure(void)
+{
+	/* Use a flag to protect the shared irq handler initialization. */
+	static bool init_irq = true;
+
+	if (init_irq) {
+		init_irq = false;
+		IRQ_CONNECT(DT_INST_IRQN(0),
+			    DT_INST_IRQ(0, priority),
+			    adc_gd32_shared_irq_handler,
+			    NULL, 0);
+		irq_enable(DT_INST_IRQN(0));
+	}
+}
```

Comment:
The same way to use to create `adc_gd32_config_##n` you can add the IRQ handle. This way you can have only 1 test each time IRQ is activated when have multiple instances.

Reply:
I'm trying to implement it, but failed. Any detailed explanation? Thanks!

Reply:
I was wondering if something like below could work:

```diff
diff --git a/drivers/adc/adc_gd32.c b/drivers/adc/adc_gd32.c
index 3d45df6acc..5590bc6c04 100644
--- a/drivers/adc/adc_gd32.c
+++ b/drivers/adc/adc_gd32.c
@@ -92,6 +92,10 @@ static void adc_gd32_isr(const struct device *dev)
        struct adc_gd32_data *data = dev->data;
        const struct adc_gd32_config *cfg = dev->config;
 
+       if (!(ADC_CTL0(cfg->reg) & ADC_CTL0_EOCIE)) {
+               return;
+       }
+
        if (ADC_STAT(cfg->reg) & ADC_STAT_EOC) {
                *data->buffer++ = ADC_RDATA(cfg->reg);
 
@@ -344,34 +348,6 @@ static int adc_gd32_init(const struct device *dev)
        return 0;
 }
 
-#define HANDLE_IRQS(idx)                                                       \
-       static const struct device *dev_##idx = DEVICE_DT_INST_GET(idx);        \
-       const struct adc_gd32_config *cfg_##idx = dev_##idx->config;            \
-                                                                               \
-       if (ADC_CTL0(cfg_##idx->reg) & ADC_CTL0_EOCIE) {                        \
-               adc_gd32_isr(dev_##idx);                                        \
-       }
-
-static void adc_gd32_shared_irq_handler(void)
-{
-       DT_INST_FOREACH_STATUS_OKAY(HANDLE_IRQS);
-}
-
-static void adc_gd32_shared_irq_configure(void)
-{
-       /* Use a flag to protect the shared irq handler initialization. */
-       static bool init_irq = true;
-
-       if (init_irq) {
-               init_irq = false;
-               IRQ_CONNECT(DT_INST_IRQN(0),
-                           DT_INST_IRQ(0, priority),
-                           adc_gd32_shared_irq_handler,
-                           NULL, 0);
-               irq_enable(DT_INST_IRQN(0));
-       }
-}
-
 #ifdef CONFIG_SOC_SERIES_GD32F3X0
 #define ADC_CLOCK_SOURCE(n)                                                    \
        .rcu_clock_source = DT_INST_PROP(n, rcu_periph_clock)
@@ -381,6 +357,15 @@ static void adc_gd32_shared_irq_configure(void)
 
 #define ADC_GD32_INIT(n)                                                       \
        PINCTRL_DT_INST_DEFINE(n);                                              \
+       static void adc##n##_gd32_irq_config(const struct device *dev)          \
+       {                                                                       \
+               IRQ_CONNECT(DT_INST_IRQN(n),                                    \
+                           DT_INST_IRQ(n, priority),                           \
+                           adc_gd32_isr,                                       \
+                           DEVICE_DT_INST_GET(n),                              \
+                           0);                                                 \
+               irq_enable(DT_INST_IRQN(n));                                    \
+       }                                                                       \
        static struct adc_gd32_data adc_gd32_data_##n = {                       \
                ADC_CONTEXT_INIT_TIMER(adc_gd32_data_##n, ctx),                 \
                ADC_CONTEXT_INIT_LOCK(adc_gd32_data_##n, ctx),                  \
@@ -391,7 +376,7 @@ static void adc_gd32_shared_irq_configure(void)
                .rcu_periph_clock = DT_INST_PROP(n, rcu_periph_clock),          \
                .channels = DT_INST_PROP(n, channels),                          \
                .pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(n),                      \
-               .irq_config_func = adc_gd32_shared_irq_configure,               \
+               .irq_config_func = adc##n##_gd32_irq_config,                    \
                ADC_CLOCK_SOURCE(n)                                             \
        };                                                                      \
        DEVICE_DT_INST_DEFINE(n,                                                \
```

Reply:
Normally, GD32 ADC sharing the interrupt between each other.

The main problem is how to get ADC0...ADCn device pointers. Without device pointer, we can not dispatch interrupt to each GD32 ADC.

Reply:
Refer from [Zephyr interrupt document](https://docs.zephyrproject.org/latest/reference/kernel/other/interrupts.html).
> Only a single ISR can be associated with a specific IRQ at any given time.